examples/gno.land/r/demo/gnolotto_factory/gnolotto_factory.gno

```diff
@@ -0,0 +1,239 @@
+func init() {
+	lotteries = avl.NewTree()
+}
+
+// Creates a new lottery, only callable by admin.
+func CreateLottery(drawTime int64, prizePool int64) (int, string) {
+	sentCoins := std.GetOrigSend()
+	amount := sentCoins.AmountOf("ugnot")
+	banker := std.GetBanker(std.BankerTypeRealmSend)
+	send := std.Coins{{"ugnot", int64(amount)}}
+
+	if prizePool != amount {
+		banker.SendCoins(std.GetOrigPkgAddr(), std.GetOrigCaller(), send)
+		return -1, "Prize pool must match the transaction value"
+	}
+
+	if drawTime < time.Now().Unix() {
+		banker.SendCoins(std.GetOrigPkgAddr(), std.GetOrigCaller(), send)
+		return -1, "Invalid draw time"
+	}
+
+	if std.GetOrigCaller() != admin {
+		banker.SendCoins(std.GetOrigPkgAddr(), std.GetOrigCaller(), send)
+		return -1, "Only the admin can create a lottery"
+	}
+
+	lotteryID := lotteries.Size()
+	lottery := gnolotto.NewLottery(time.Unix(drawTime, 0), prizePool)
+
+	lotteries.Set(ufmt.Sprintf("%d", lotteryID), lottery)
+	return lotteryID, "Lottery created successfully"
+}
+
+// Buy ticket for a specific lottery.
+func BuyTicket(lotteryID int, numbersStr string) (int, string) {
+	sentCoins := std.GetOrigSend()
+	amount := sentCoins.AmountOf("ugnot")
+	banker := std.GetBanker(std.BankerTypeRealmSend)
+	send := std.Coins{{"ugnot", int64(amount)}}
+
+	id := ufmt.Sprintf("%d", lotteryID)
+	lotteryRaw, exists := lotteries.Get(id)
+	if !exists {
+		banker.SendCoins(std.GetOrigPkgAddr(), std.GetOrigCaller(), send)
+		return -1, "Lottery not found"
+	}
+
+	// Convert string to slice of integers.
+	numbersSlice := strings.Split(numbersStr, ",")
+	numbers := make([]int, len(numbersSlice))
+	for i, numStr := range numbersSlice {
+		num, err := strconv.Atoi(numStr)
+		if err != nil {
+			banker.SendCoins(std.GetOrigPkgAddr(), std.GetOrigCaller(), send)
+			panic("Invalid number: " + err.Error())
+		}
+		numbers[i] = num
+	}
+
+	//Verify if the amount sent is equal to the ticket price.
+	if amount != 10 {
+		banker.SendCoins(std.GetOrigPkgAddr(), std.GetOrigCaller(), send)
+		return -1, "Ticket price must be 10 UGNOT"
+	}
+
+	// Verify if the numbers are unique.
+	uniqueNumbers := make(map[int]bool)
+	for _, num := range numbers {
+		if uniqueNumbers[num] {
+			banker.SendCoins(std.GetOrigPkgAddr(), std.GetOrigCaller(), send)
+			return -1, "Numbers must be unique"
+		}
+		uniqueNumbers[num] = true
+	}
+
+	l, _ := lotteryRaw.(*gnolotto.Lottery)
+
+	if time.Now().Unix() > l.DrawTime.Unix() {
+		banker.SendCoins(std.GetOrigPkgAddr(), std.GetOrigCaller(), send)
+		return -1, "This lottery has already ended"
+	}
+
+	if len(numbers) > 5 || len(numbers) < 5 {
+		banker.SendCoins(std.GetOrigPkgAddr(), std.GetOrigCaller(), send)
+		return -1, "You must select exactly 5 numbers"
+	}
+
+	for _, num := range numbers {
+		if num > 15 || num < 1 {
+			banker.SendCoins(std.GetOrigPkgAddr(), std.GetOrigCaller(), send)
+			return -1, "Invalid number, select number range from 1 to 15"
+		}
+	}
+
+	caller := std.GetOrigCaller()
```

Comment:
You're using std.getOrigCaller() many times in this function. How about just calling it once at the top of the function, and using `caller` every other time? Blockchain code uses gas to execute, which is why it's important to optimize it.

Reply:
following the modification of errors by adding `panic` I only use `std.getOrigCaller()` once in this function